### ISSUE TITLE
Inference with CPU only

### ISSUE THREAD

**Issue by DangChuong-DC** [csuhan/ReDet#131]
Hello @csuhan,
Thank you very much for sharing your great work!
I am wondering if ReDet is supporting inference with only CPU?

If it is supporting, how should we do it?
If not, will you consider to support it in the future?

**Comment by csuhan** [comment]
ReDet supports cpu inference.
Try to set `device='cpu'` in https://github.com/csuhan/ReDet/blob/0b9addf3c2734659fd6ffc7824f2e659fde4419c/demo_large_image.py#L91

**Comment by DangChuong-DC** [comment]
Hi @csuhan,
Thank you very much for your prompt response!
Running inference with GPU(s) works totally fine.
However, when I tried your suggestion for running inference with cpu only, there is an error, it seems like some piece of the model is still in 'cuda:0'
I provided the screen shot of the error, do you have any thoughts on this?
![Screenshot from 2021-12-16 13-33-31](https://user-images.githubusercontent.com/40515271/146309042-06de25ff-af7f-474d-937b-a4b84284d823.png)

**Comment by jigongbao** [comment]
@csuhan  i have the similar prombles too , i try to run inference with cuda:4 but there is the same error.
![2~E7A$W3L0K1Y%)N5V0$_TV](https://user-images.githubusercontent.com/96114577/146327998-d220baf2-6099-4d2a-bffe-e77bfff3aa57.png)

**Comment by DangChuong-DC** [comment]
@jigongbao your problem can be solved by changing the code in `mmdet/models/anchor_head.py` as below, I guess but not sure 
![Screenshot from 2021-12-16 17-33-31](https://user-images.githubusercontent.com/40515271/146336242-8d94b6ca-28fc-4195-ab99-ea604ad6935b.png)

For my problem with CPU inference, I modified a bit and it comes to this error of those file in `mmdet/models/ops`
![Screenshot from 2021-12-16 17-39-38](https://user-images.githubusercontent.com/40515271/146337139-03f207ee-935b-49c7-abd8-5ce20725deba.png)

**Comment by csuhan** [comment]
@DangChuong-DC  You are right. The codebase of ReDet is a rather early version of mmdet, and some ops do not suport running on cpu.
I suggest to use corresponding ops in mmcv which support cpu inference.
For example, 
RoIAlign: https://github.com/open-mmlab/mmcv/blob/master/mmcv/ops/roi_align.py
RoIAlign Rotated: https://github.com/open-mmlab/mmcv/blob/master/mmcv/ops/roi_align_rotated.py

But for our RiRoIAlign, I have no time to implement the cpu version. If you still want to use it, please refer to:
https://github.com/csuhan/ReDet/blob/master/mmdet/ops/riroi_align/src/riroi_align_kernel.cu
https://github.com/open-mmlab/mmcv/blob/master/mmcv/ops/csrc/pytorch/cpu/roi_align_rotated.cpp
I think it would not be hard to implement the cpu version.

**Comment by jigongbao** [comment]
@DangChuong-DC   i tried , but there is a new error.... but thanks 
![image](https://user-images.githubusercontent.com/96114577/146381503-d297873f-caa1-4405-b536-77126c2c8b71.png)

**Comment by csuhan** [comment]
@jigongbao , try `nms_cpu` instead.
https://github.com/csuhan/ReDet/blob/0b9addf3c2734659fd6ffc7824f2e659fde4419c/mmdet/ops/nms/nms_wrapper.py#L48-L51

**Comment by jigongbao** [comment]
@csuhan   thanks
![image](https://user-images.githubusercontent.com/96114577/146493546-3fe9eb44-cfe7-4afb-92f1-f4cca51a9f2c.png)
![image](https://user-images.githubusercontent.com/96114577/146493568-09cd02cd-2696-4001-86f1-f7d88c9bc801.png)
![image](https://user-images.githubusercontent.com/96114577/146493750-05bfe250-267e-4cc9-94ea-1e7a27980b2d.png)
i try to change some code in other place and it works correctly.

**Comment by DangChuong-DC** [comment]
@csuhan Thanks a lot for your explanations, I will have a look on it
